File: ruby/blackjack/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
      when "ace"   then 11
      when "two"   then 2
      when "three" then 3
      when "four"  then 4
      when "five"  then 5
      when "six"   then 6
      when "seven" then 7
      when "eight" then 8
      when "nine"  then 9
      when "ten", "jack", "queen", "king" then 10
      else 0
    end
  end

  def self.card_range(card1, card2)
    case self.parse_card(card1) + self.parse_card(card2)
      when  4..11 then "low"
      when 12..16 then "mid"
      when 17..20 then "high"
      when 21     then "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    r = self.card_range(card1, card2)
    dc = self.parse_card(dealer_card)
    if card1 == 'ace' && card2 == 'ace' then 'P'
    elsif r == 'blackjack' then ([10,11].include?(dc)) ? 'S' : 'W'
    elsif r == 'high' then 'S'
    elsif r == 'mid' && dc < 7 then 'S'
    else 'H'
    end
  end
end

File: go/blackjack/blackjack.go
package blackjack

// Ternary function mimics the ternary operator
func Ternary[T any](condition bool, trueVal, falseVal T) T {
	if condition {
		return trueVal
	}
	return falseVal
}

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "ace":
		return 11
	case "king":
		return 10
	case "queen":
		return 10
	case "jack":
		return 10
	case "ten":
		return 10
	case "nine":
		return 9
	case "eight":
		return 8
	case "seven":
		return 7
	case "six":
		return 6
	case "five":
		return 5
	case "four":
		return 4
	case "three":
		return 3
	case "two":
		return 2
	case "one":
		return 1
	default:
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	v1 := ParseCard(card1)
	v2 := ParseCard(card2)
	dv := ParseCard(dealerCard)
	total := v1 + v2
	switch {
	case total == 21:
		return Ternary(dv < 10, "W", "S")
	case total == 22:
		return "P"
	case total >= 17:
		return "S"
	case total >= 12:
		return Ternary(dv < 7, "S", "H")
	default:
		return "H"
	}
}
